examples/step-97/step-97.cc

```diff
@@ -0,0 +1,2125 @@
+    , dofs_Hcurl(scratch_data.dofs_Hcurl)
+  {}
+
+  // This function assembles a fraction of the system matrix and the system
+  // right-hand side related to a single cell. These fractions are
+  // `copy_data.cell_matrix` and `copy_data.cell_rhs`. They are copied into
+  // to the system matrix, $A_{ij}$, and the right-hand side, $b_i$, by the
+  // function `Solver::copy_local_to_global()`.
+  void Solver::system_matrix_local(const IteratorPair  &IP,
+                                   AssemblyScratchData &scratch_data,
+                                   AssemblyCopyData    &copy_data)
+  {
+    // First we reinitialize the matrices and vectors related to the current
+    // cell, update the FE values, and compute the gradients of the input vector
+    // field.
+    copy_data.cell_matrix.reinit(scratch_data.dofs_per_cell,
+                                 scratch_data.dofs_per_cell);
+
+    copy_data.cell_rhs.reinit(scratch_data.dofs_per_cell);
+
+    copy_data.local_dof_indices.resize(scratch_data.dofs_per_cell);
+
+    scratch_data.fe_values_Hdiv.reinit(std::get<0>(*IP));
+    scratch_data.fe_values_Hcurl.reinit(std::get<1>(*IP));
+
+    scratch_data.fe_values_Hcurl.get_function_gradients(
+      scratch_data.dofs_Hcurl, scratch_data.vector_gradients);
+
+    // Second, we compute the components of the cell matrix and cell right-hand
+    // side. The labels of the integrals are the same as in the introduction to
+    // this tutorial.
+    for (unsigned int q_index = 0; q_index < scratch_data.n_q_points; ++q_index)
+      {
+        for (unsigned int i = 0; i < scratch_data.dofs_per_cell; ++i)
+          {
+            for (unsigned int j = 0; j < scratch_data.dofs_per_cell; ++j)
+              {
+                copy_data.cell_matrix(i, j) += // Integral I_a
+                  scratch_data.fe_values_Hdiv[scratch_data.ve].value(i,
+                                                                     q_index) *
+                  scratch_data.fe_values_Hdiv[scratch_data.ve].value(j,
+                                                                     q_index) *
+                  scratch_data.fe_values_Hdiv.JxW(q_index);
+              }
+
+            // The variable `curl_vec_in_Hcurl` denotes the curl of the input
+            // vector field, $\vec{\nabla} \times \vec{T}$ or
+            // $\vec{\nabla} \times \vec{A}$, depending on the context.
+            scratch_data.curl_vec_in_Hcurl[0] =
+              scratch_data.vector_gradients.at(q_index).at(2)[1] -
+              scratch_data.vector_gradients.at(q_index).at(1)[2];
+
+            scratch_data.curl_vec_in_Hcurl[1] =
+              scratch_data.vector_gradients.at(q_index).at(0)[2] -
+              scratch_data.vector_gradients.at(q_index).at(2)[0];
+
+            scratch_data.curl_vec_in_Hcurl[2] =
+              scratch_data.vector_gradients.at(q_index).at(1)[0] -
+              scratch_data.vector_gradients.at(q_index).at(0)[1];
```

Comment:
You're extracting the curl from the tensor gradients here. Couldn't you get these directly if you replaced `scratch_data.fe_values_Hcurl.get_function_gradients` by `scratch_data.fe_values_Hcurl[ve].get_function_curls`?

Separately: Instead of `.at(q_index).at(...)`, let's use `[q_index][...]` as you do everywhere else (I believe).